src/EditorFeatures/Core/Implementation/SplitComment/SplitCommentCommandHandler.cs

```diff
@@ -0,0 +1,218 @@
+                return SplitCommentAsync(textView, subjectBuffer, document, position, cancellationToken).WaitAndGetResult(cancellationToken);
+            }
+        }
+
+        private static bool LineProbablyContainsComment(ISplitCommentService service, ITextSnapshotLine line, int caretPosition)
+        {
+            var commentStart = service.CommentStart;
+
+            var end = Math.Max(caretPosition, line.Length);
+            for (var i = 0; i < end; i++)
+            {
+                if (MatchesCommentStart(line, commentStart, i))
+                    return true;
+            }
+
+            return false;
+        }
+
+        private static bool MatchesCommentStart(ITextSnapshotLine line, string commentStart, int index)
+        {
+            var lineStart = line.Start;
+            for (var c = 0; c < commentStart.Length; c++)
+            {
+                if (lineStart.Position + index >= line.Snapshot.Length)
+                    return false;
+
+                if (line.Snapshot[lineStart + index] != commentStart[c])
+                    return false;
+            }
+
+            return true;
+        }
+
+        private async Task<bool> SplitCommentAsync(
+            ITextView textView, ITextBuffer subjectBuffer, Document document, int position, CancellationToken cancellationToken)
+        {
+            var options = await document.GetOptionsAsync(cancellationToken).ConfigureAwait(false);
+            var enabled = options.GetOption(SplitCommentOptions.Enabled);
+            if (!enabled)
+                return false;
+
+            var root = await document.GetRequiredSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+            var syntaxKinds = document.GetRequiredLanguageService<ISyntaxKindsService>();
+            var trivia = root.FindTrivia(position);
+            if (syntaxKinds.SingleLineCommentTrivia != trivia.RawKind)
+                return false;
+
+            var splitCommentService = document.GetRequiredLanguageService<ISplitCommentService>();
+
+            // if the user hits enter at `/$$/` we don't want to consider this a comment continuation.
+            if (position < (trivia.SpanStart + splitCommentService.CommentStart.Length))
+                return false;
+
+            if (!splitCommentService.IsAllowed(root, trivia))
+                return false;
+
+            var textSnapshot = subjectBuffer.CurrentSnapshot;
+            var triviaLine = textSnapshot.GetLineFromPosition(trivia.SpanStart);
+
+            using var transaction = CaretPreservingEditTransaction.TryCreate(
+                EditorFeaturesResources.Split_comment, textView, _undoHistoryRegistry, _editorOperationsFactoryService);
+
+            subjectBuffer.Replace(
+                GetReplacementSpan(triviaLine, position),
+                GetReplacementText(textView, document, options, triviaLine, trivia, position));
+
+            transaction.Complete();
```

Comment:
❔ Does this need to be on the main thread?

Reply:
that is a good question.  I observe no bad behavior doing things this way... but it is somewhat suspect.  Wondering if @jasonmalinowski or @olegtk  might know...  i can rejigger this though just to be safe in the meantime.